Feature/fermion matrix/hubbard exp

Marcel-Rodekamp opened the pull request:
Fermion matrix as of #37. The logDet tests fail, but we should still merge. A fix for these is discussed in #36 & #43.

Marcel-Rodekamp committed:
Added matrix multiplication as NSL::Tensor::mat_mul

Marcel-Rodekamp committed:
Added an implementation for the Fermion Matrix class: Hubbard Model Exponential discretization

Marcel-Rodekamp committed:
Added Tests for the Fermion Matrix: Hubbard Model Exponential Discretization. The logDet tests are considered to fail, for details follow the discussion in #36 & #43

keshvituteja reviewed (commented):
Everything looks good to me!

evanberkowitz commented on Tests/CMakeLists.txt:
```diff
@@ -1,3 +1,4 @@
+add_subdirectory(FermionMatrix/)
```
Does putting this first mean that the tests will run first?  That seems counter intuitive---I would have guessed: first test Tensor, then LinAlg, then FermionMatrix.  Or is there a reason to put this first?

Marcel-Rodekamp replied:
Solved with commit 7801f46bde87d495cc2b6f10cb473935a82aa279.

evanberkowitz commented on src/NSL/FermionMatrix/Impl/hubbardExp.hpp:
```diff
@@ -0,0 +1,91 @@
+    //!
+    /*!
+     *
+     * \todo: SpatialLattice accepts only real type template arguments
```
What?  Why?  There are compiling examples where SpatialLattice takes complex arguments.

evanberkowitz commented on Tests/test.hpp:
```diff
@@ -85,6 +85,38 @@ bool compare_integer(Tint a, Tint b, Tint prec = 0){
+    return std::fabs(x-y) <= std::numeric_limits<Type>::epsilon() * max * ulp
+        // unless the result is subnormal
+        || std::fabs(x-y) < std::numeric_limits<Type>::min();
+}
```
This is lifted almost verbatim from cppreference.  Their licenses are CC-BY-SA-3.0 and GFDL.  I'm not sure you can promote GFDL --> GPL.  So if you want to go with GPL I'm not sure this code can be merged.

evanberkowitz replied:
In fact [the GFDL wikipedia article](https://en.wikipedia.org/wiki/GNU_Free_Documentation_License) suggests GFDL is incompatible with GPL in both directions.

Marcel-Rodekamp replied:
So I scanned through the FAQ from [cppreference.cpp](https://en.cppreference.com/w/Cppreference:FAQ). Here they state:
```
The content is licensed under[ Creative Commons Attribution-Sharealike 3.0 Unported License]
(https://en.cppreference.com/w/Cppreference:Copyright/CC-BY-SA) (CC-BY-SA) 
and by the GNU Free Documentation License (GFDL) (unversioned, with no invariant sections, 
front-cover texts, or back-cover texts). That means that you can use this site in 
almost any way you like, including mirroring, copying, translating, etc. All we would 
ask is to provide link back to cppreference.com so that people know where to get the 
most up-to-date content. In addition to that, any modified content should be released 
under an equivalent license so that everyone could benefit from the modified versions.
```
So I would like to raise the questions: 
1. When is the changed code so close that it counts under this restriction at all
2. What licenses are equivalent or do we plan a equivalent license anyway and we wouldn't need to bother.

evanberkowitz replied:
It's the "equivalent license" part that is the issue!

evanberkowitz replied:
It looks like [we can license our version as CC-BY-SA-4.0](https://creativecommons.org/share-your-work/licensing-considerations/compatible-licenses).  Since CC-BY-SA-4.0 is GPLv3 compatible (same link), if we also decide to license NSL as GPLv3 that should be fine.

evanberkowitz replied:
Also, we now use a different strategy where we specify the number of decimal digits.  I think this is "different enough" to not worry.

evanberkowitz commented on src/NSL/FermionMatrix/Impl/hubbardExp.tpp:
```diff
@@ -0,0 +1,124 @@
+        + this->Mdagger(psi) 
+        + NSL::LinAlg::mat_vec(
+            (this->Lat->exp_hopping_matrix(/*delta=(beta/Nt) */delta_))*
+            (this->Lat->exp_hopping_matrix(/*delta=(beta/Nt) */delta_)), 
```
Shouldn't we just ask for `exp_hopping_matrix(2*delta_)`?

evanberkowitz replied:
Even if not, this does (mat*mat) * vec, rather than mat * (mat * vec), so scales like Nx^3 rather than Nx^2.

evanberkowitz replied:
ALSO, does the `*` do the matrix multiplication desired?

evanberkowitz commented on src/NSL/FermionMatrix/Impl/hubbardExp.tpp:
```diff
@@ -0,0 +1,124 @@
+    //MMdagger(psi) = M(psi) + Mdagger(psi) + exp(-i*phi) * (exp_hopping_mtrix^2 x exp(-i*phi)* psi) - psi
+    NSL::Tensor<Type> out= this->M(psi) + this->Mdagger(psi) + (NSL::LinAlg::adjoint(this->phiExp_) *
+        NSL::LinAlg::mat_vec((this->Lat->exp_hopping_matrix(/*delta=(beta/Nt) */delta_))*
+         (this->Lat->exp_hopping_matrix(/*delta=(beta/Nt) */delta_)),
```
Should this really be `*` and not matrix multiplication?

Probably better to just call on `(2*delta_)`.

keshvituteja replied:
[4ff21a7](https://github.com/Marcel-Rodekamp/NSL/pull/65/commits/4ff21a764c9b6957d976355035797040c3ce6d47) resolves this

evanberkowitz reviewed (changes requested):
I have MANY questions about this PR!

evanberkowitz commented:
This actually doesn't compile for me.  While building
```
[  9%] Building CXX object Tests/FermionMatrix/CMakeFiles/test_fermionMatrixHubbardExp.dir/test_fermionMatrixHubbardExp.cpp.o
```
I get 
```
nsl/src/NSL/FermionMatrix/Impl/hubbardExp.hpp:83:5: error: missing 'typename' prior to dependent type name 'NSL::RT_extractor<Type>::value_type'
    NSL::RT_extractor<Type>::value_type delta_;
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    typename
```

Marcel-Rodekamp committed:
Updated documentary of `almost_equal`

keshvituteja committed:
Spatial Lattice object also accepts complex arguments now

keshvituteja committed:
Spatial Lattice object also accepts complex arguments now

keshvituteja committed:
replaced exp_hopping_matrix(delta)*exp_hopping_matrix(delta) with exp_hopping_matrix(2*delta) since it's a mat_vec instead of elementwise multiplication

keshvituteja committed:
Spatial Lattice object also accepts complex arguments now

keshvituteja committed:
Merge branch 'feature/fermionMatrix/HubbardExp' of https://github.com/Marcel-Rodekamp/NSL into feature/fermionMatrix/HubbardExp
Spatial Lattice Object accepts complex types now

evanberkowitz commented:
I tried compiling and still ran into the issue regarding the the type of delta.  I changed the declaration of the private delta_ to `typename NSL::RT_extractor<Type>::value_type delta_;`.  That fixed some issues.  However, now I face the following compilation problem

```
In file included from /Users/evanberkowitz/physics/hubbard/nsl/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:1:
In file included from /Users/evanberkowitz/physics/hubbard/nsl/Tests/FermionMatrix/../test.hpp:5:
In file included from /Users/evanberkowitz/physics/hubbard/nsl/src/NSL/NSL.hpp:14:
In file included from /Users/evanberkowitz/physics/hubbard/nsl/src/NSL/FermionMatrix/../Tensor.hpp:5:
In file included from /Users/evanberkowitz/physics/hubbard/nsl/src/NSL/Tensor/Factory/../tensor.hpp:15:
/Users/evanberkowitz/physics/hubbard/nsl/src/NSL/Tensor/Impl/randomAccess.tpp:32:17: error: no viable constructor or deduction guide for deduction of template arguments of 'initializer_list'
                std::initializer_list{torch::indexing::TensorIndex(indexer)...}
                ^
/Users/evanberkowitz/physics/hubbard/nsl/src/NSL/FermionMatrix/Impl/hubbardExp.tpp:106:11: note: in instantiation of function template specialization 'NSL::TensorImpl::TensorRandomAccess<c10::complex<float>>::operator()<int, NSL::Slice, NSL::Slice>' requested here
    out = prod(Nt-1, NSL::Slice(), NSL::Slice());
          ^
/Users/evanberkowitz/physics/hubbard/nsl/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:107:49: note: in instantiation of member function 'NSL::FermionMatrix::FermionMatrixHubbardExp<c10::complex<float>>::logDetM' requested here
    NSL::FermionMatrix::FermionMatrixHubbardExp M(&r,phi,2.0);
                                                ^
/Users/evanberkowitz/physics/hubbard/nsl/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:33:5: note: in instantiation of function template specialization 'test_fermionMatrixHubbardExp_M<float>' requested here
    test_fermionMatrixHubbardExp_M<TestType>(size_0, size_1);```

Marcel-Rodekamp committed:
Changed order of tests, the fermion matrix tests now run after the tensor and lattice ones

Marcel-Rodekamp commented:
> This actually doesn't compile for me. While building
> 
> ```
> [  9%] Building CXX object Tests/FermionMatrix/CMakeFiles/test_fermionMatrixHubbardExp.dir/test_fermionMatrixHubbardExp.cpp.o
> ```
> 
> I get
> 
> ```
> nsl/src/NSL/FermionMatrix/Impl/hubbardExp.hpp:83:5: error: missing 'typename' prior to dependent type name 'NSL::RT_extractor<Type>::value_type'
>     NSL::RT_extractor<Type>::value_type delta_;
>     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     typename
> ```

Mhh, why does the `detla_` variable needs to be real here? didn't we allow for complex values in the Lattice classes?

evanberkowitz commented:
Indeed; it's fine to just let it be `Type`.  But even then I get the compilation error above.

Marcel-Rodekamp commented:
> I tried compiling and still ran into the issue regarding the the type of delta. I changed the declaration of the private delta_ to `typename NSL::RT_extractor<Type>::value_type delta_;`. That fixed some issues. However, now I face the following compilation problem
> 
> ```
> In file included from /Users/evanberkowitz/physics/hubbard/nsl/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:1:
> In file included from /Users/evanberkowitz/physics/hubbard/nsl/Tests/FermionMatrix/../test.hpp:5:
> In file included from /Users/evanberkowitz/physics/hubbard/nsl/src/NSL/NSL.hpp:14:
> In file included from /Users/evanberkowitz/physics/hubbard/nsl/src/NSL/FermionMatrix/../Tensor.hpp:5:
> In file included from /Users/evanberkowitz/physics/hubbard/nsl/src/NSL/Tensor/Factory/../tensor.hpp:15:
> /Users/evanberkowitz/physics/hubbard/nsl/src/NSL/Tensor/Impl/randomAccess.tpp:32:17: error: no viable constructor or deduction guide for deduction of template arguments of 'initializer_list'
>                 std::initializer_list{torch::indexing::TensorIndex(indexer)...}
>                 ^
> /Users/evanberkowitz/physics/hubbard/nsl/src/NSL/FermionMatrix/Impl/hubbardExp.tpp:106:11: note: in instantiation of function template specialization 'NSL::TensorImpl::TensorRandomAccess<c10::complex<float>>::operator()<int, NSL::Slice, NSL::Slice>' requested here
>     out = prod(Nt-1, NSL::Slice(), NSL::Slice());
>           ^
> /Users/evanberkowitz/physics/hubbard/nsl/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:107:49: note: in instantiation of member function 'NSL::FermionMatrix::FermionMatrixHubbardExp<c10::complex<float>>::logDetM' requested here
>     NSL::FermionMatrix::FermionMatrixHubbardExp M(&r,phi,2.0);
>                                                 ^
> /Users/evanberkowitz/physics/hubbard/nsl/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:33:5: note: in instantiation of function template specialization 'test_fermionMatrixHubbardExp_M<float>' requested here
>     test_fermionMatrixHubbardExp_M<TestType>(size_0, size_1);```
> ```

 

This should be fixed with #68. We should Solve that PR first and then rebase from devel.

keshvituteja commented:
```
> I tried compiling and still ran into the issue regarding the the type of delta. I changed the declaration of the private delta_ to `typename NSL::RT_extractor<Type>::value_type delta_;`. That fixed some issues. However, now I face the following compilation problem
> 
> ```
> In file included from /Users/evanberkowitz/physics/hubbard/nsl/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:1:
> In file included from /Users/evanberkowitz/physics/hubbard/nsl/Tests/FermionMatrix/../test.hpp:5:
> In file included from /Users/evanberkowitz/physics/hubbard/nsl/src/NSL/NSL.hpp:14:
> In file included from /Users/evanberkowitz/physics/hubbard/nsl/src/NSL/FermionMatrix/../Tensor.hpp:5:
> In file included from /Users/evanberkowitz/physics/hubbard/nsl/src/NSL/Tensor/Factory/../tensor.hpp:15:
> /Users/evanberkowitz/physics/hubbard/nsl/src/NSL/Tensor/Impl/randomAccess.tpp:32:17: error: no viable constructor or deduction guide for deduction of template arguments of 'initializer_list'
>                 std::initializer_list{torch::indexing::TensorIndex(indexer)...}
>                 ^
> /Users/evanberkowitz/physics/hubbard/nsl/src/NSL/FermionMatrix/Impl/hubbardExp.tpp:106:11: note: in instantiation of function template specialization 'NSL::TensorImpl::TensorRandomAccess<c10::complex<float>>::operator()<int, NSL::Slice, NSL::Slice>' requested here
>     out = prod(Nt-1, NSL::Slice(), NSL::Slice());
>           ^
> /Users/evanberkowitz/physics/hubbard/nsl/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:107:49: note: in instantiation of member function 'NSL::FermionMatrix::FermionMatrixHubbardExp<c10::complex<float>>::logDetM' requested here
>     NSL::FermionMatrix::FermionMatrixHubbardExp M(&r,phi,2.0);
>                                                 ^
> /Users/evanberkowitz/physics/hubbard/nsl/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:33:5: note: in instantiation of function template specialization 'test_fermionMatrixHubbardExp_M<float>' requested here
>     test_fermionMatrixHubbardExp_M<TestType>(size_0, size_1);```
> ```
```

I am not sure what's causing the error, I could give a description about the changes I committed, if that helps in any way:

- commit [a64bfe1](https://github.com/Marcel-Rodekamp/NSL/pull/65/commits/a64bfe152b5a5ebbc4f2122dd0dd44a0e90485d2), changes the type of the Lat object from `NSL::Lattice::SpatialLattice<typename NSL::RT_extractor<Type>::value_type>* Lat;` to `NSL::Lattice::SpatialLattice<Type>* Lat;` in `FermionMatrix` class (and also changes the Type of Lat object in constructor) in src/NSL/FermionMatrix/fermionMatrix.hpp

- commit [e66957d](https://github.com/Marcel-Rodekamp/NSL/pull/65/commits/e66957de9290cce0c7ddec564ee35e4e7d0b8ff8) changes the arguments in the constructor of `FermionMatrixHubbardExp` class from `FermionMatrixHubbardExp(NSL::Lattice::SpatialLattice<typename NSL::RT_extractor<Type>::value_type> * lat, const NSL::Tensor<Type> &phi, double beta = 1.0 )` to `FermionMatrixHubbardExp(NSL::Lattice::SpatialLattice<Type> * lat, const NSL::Tensor<Type> &phi, double beta = 1.0 )` in src/NSL/FermionMatrix/Impl/hubbardExp.hpp

- commit [7bdb448](https://github.com/Marcel-Rodekamp/NSL/pull/65/commits/7bdb448f6b8e4496fe9764deeee00b0e5e47fa01) just changes the type of the ring object from  `NSL::Lattice::Ring<Type> r(size1);` to `NSL::Lattice::Ring<NSL::complex<Type>> r(size1);` for all the test cases in Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp (and also replaces `(r.exp_hopping_matrix(delta))*(r.exp_hopping_matrix(delta))` with `(r.exp_hopping_matrix(2*delta))` )

- And commit [4ff21a7](https://github.com/Marcel-Rodekamp/NSL/pull/65/commits/4ff21a764c9b6957d976355035797040c3ce6d47) just fixes the same bug of `this->Lat->exp_hopping_matrix(2*delta_) ` in src/NSL/FermionMatrix/Impl/hubbardExp.tpp

Marcel-Rodekamp committed:
Added concept: isDerived

Marcel-Rodekamp committed:
Added template LatticeType to FermionMatrix

Marcel-Rodekamp committed:
Added template LatticeType to HubbardExp, changed name FermionMatrixHubbardExp to HubbardExp

Marcel-Rodekamp commented:
> ```
> > I tried compiling and still ran into the issue regarding the the type of delta. I changed the declaration of the private delta_ to `typename NSL::RT_extractor<Type>::value_type delta_;`. That fixed some issues. However, now I face the following compilation problem
> > 
> > ```
> > In file included from /Users/evanberkowitz/physics/hubbard/nsl/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:1:
> > In file included from /Users/evanberkowitz/physics/hubbard/nsl/Tests/FermionMatrix/../test.hpp:5:
> > In file included from /Users/evanberkowitz/physics/hubbard/nsl/src/NSL/NSL.hpp:14:
> > In file included from /Users/evanberkowitz/physics/hubbard/nsl/src/NSL/FermionMatrix/../Tensor.hpp:5:
> > In file included from /Users/evanberkowitz/physics/hubbard/nsl/src/NSL/Tensor/Factory/../tensor.hpp:15:
> > /Users/evanberkowitz/physics/hubbard/nsl/src/NSL/Tensor/Impl/randomAccess.tpp:32:17: error: no viable constructor or deduction guide for deduction of template arguments of 'initializer_list'
> >                 std::initializer_list{torch::indexing::TensorIndex(indexer)...}
> >                 ^
> > /Users/evanberkowitz/physics/hubbard/nsl/src/NSL/FermionMatrix/Impl/hubbardExp.tpp:106:11: note: in instantiation of function template specialization 'NSL::TensorImpl::TensorRandomAccess<c10::complex<float>>::operator()<int, NSL::Slice, NSL::Slice>' requested here
> >     out = prod(Nt-1, NSL::Slice(), NSL::Slice());
> >           ^
> > /Users/evanberkowitz/physics/hubbard/nsl/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:107:49: note: in instantiation of member function 'NSL::FermionMatrix::FermionMatrixHubbardExp<c10::complex<float>>::logDetM' requested here
> >     NSL::FermionMatrix::FermionMatrixHubbardExp M(&r,phi,2.0);
> >                                                 ^
> > /Users/evanberkowitz/physics/hubbard/nsl/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:33:5: note: in instantiation of function template specialization 'test_fermionMatrixHubbardExp_M<float>' requested here
> >     test_fermionMatrixHubbardExp_M<TestType>(size_0, size_1);```
> > ```
> ```
> 
> I am not sure what's causing the error, I could give a description about the changes I committed, if that helps in any way:
> 
>     * commit [a64bfe1](https://github.com/Marcel-Rodekamp/NSL/pull/65/commits/a64bfe152b5a5ebbc4f2122dd0dd44a0e90485d2), changes the type of the Lat object from `NSL::Lattice::SpatialLattice<typename NSL::RT_extractor<Type>::value_type>* Lat;` to `NSL::Lattice::SpatialLattice<Type>* Lat;` in `FermionMatrix` class (and also changes the Type of Lat object in constructor) in src/NSL/FermionMatrix/fermionMatrix.hpp
> 
>     * commit [e66957d](https://github.com/Marcel-Rodekamp/NSL/pull/65/commits/e66957de9290cce0c7ddec564ee35e4e7d0b8ff8) changes the arguments in the constructor of `FermionMatrixHubbardExp` class from `FermionMatrixHubbardExp(NSL::Lattice::SpatialLattice<typename NSL::RT_extractor<Type>::value_type> * lat, const NSL::Tensor<Type> &phi, double beta = 1.0 )` to `FermionMatrixHubbardExp(NSL::Lattice::SpatialLattice<Type> * lat, const NSL::Tensor<Type> &phi, double beta = 1.0 )` in src/NSL/FermionMatrix/Impl/hubbardExp.hpp
> 
>     * commit [7bdb448](https://github.com/Marcel-Rodekamp/NSL/pull/65/commits/7bdb448f6b8e4496fe9764deeee00b0e5e47fa01) just changes the type of the ring object from  `NSL::Lattice::Ring<Type> r(size1);` to `NSL::Lattice::Ring<NSL::complex<Type>> r(size1);` for all the test cases in Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp (and also replaces `(r.exp_hopping_matrix(delta))*(r.exp_hopping_matrix(delta))` with `(r.exp_hopping_matrix(2*delta))` )
> 
>     * And commit [4ff21a7](https://github.com/Marcel-Rodekamp/NSL/pull/65/commits/4ff21a764c9b6957d976355035797040c3ce6d47) just fixes the same bug of `this->Lat->exp_hopping_matrix(2*delta_) ` in src/NSL/FermionMatrix/Impl/hubbardExp.tpp

This is the same what @evanberkowitz has found and is related to #68

evanberkowitz commented:
I merged #68.

evanberkowitz commented:
After merging #68, suddenly there was no longer an `NSL::FermionMatrix::FermionMatrixHubbardExp` but I assume the name has been changed to `NSL::FermionMatrix::HubbardExp`.  I tried changing that throughout, but then got errors such as
```
/Users/evanberkowitz/physics/hubbard/nsl/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:282:25: error: too few template arguments for class template 'HubbardExp'
    NSL::FermionMatrix::HubbardExp<NSL::complex<Type>> M(&ring,phi,0.1);
                        ^
/Users/evanberkowitz/physics/hubbard/nsl/src/NSL/FermionMatrix/Impl/hubbardExp.hpp:19:7: note: template is declared here
class HubbardExp : public FermionMatrix<Type,LatticeType> {
```
so it's not as straightforward a fix as I imagined.

Marcel-Rodekamp commented:
I fear there is a clash happening
along with the name change from `FermionMatrixHubbardExp` -> `HubbardExp` I changed the behaviour of the lattice class. The `HubbardExp` takes a reference and has a template which determines its type. Compare commit  [6cb5b5f](https://github.com/Marcel-Rodekamp/NSL/pull/65/commits/6cb5b5ff675b5c7f54be60ed84cb64facbcbec54) and [7ff8fb6](https://github.com/Marcel-Rodekamp/NSL/pull/65/commits/7ff8fb62f9e36c9a46b56b5c3c958850da590506).

Edit: Sorry, my mistake I forgot to add the test file... with [58d4527](https://github.com/Marcel-Rodekamp/NSL/pull/65/commits/58d452771527afbeb6447f91adb5ae7932277071) it should work

Marcel-Rodekamp committed:
Updated the Test for the change in HubbardExp

evanberkowitz commented:
OK, with that it compiles and the tests (except for the expected `float` determinant tests) pass.

Before we merge, however, I'd like to edit the tests a little bit to be not so reliant on `Ring` only.

Marcel-Rodekamp commented:
Agreed, I would like to have one more scan as well @keshvituteja is the fix with the shift already in here?

keshvituteja commented:
The shift fix is not in here, I got some errors when I tried to compile it yesterday after the meeting. Error:

```/mnt/f/NSL/src/NSL/Tensor/Impl/shift.tpp: In instantiation of ‘NSL::Tensor<Type> NSL::TensorImpl::TensorShift<Type>::shift(const size_t&, const Type&) [with Type = c10::complex<float>; NSL::size_t = long int]’:
/mnt/f/NSL/src/NSL/LinAlg/shift.tpp:23:43:   required from ‘NSL::Tensor<Type> NSL::LinAlg::shift(const NSL::Tensor<Type>&, const size_t&, const Type&) [with Type = c10::complex<float>; NSL::size_t = long int]’
/mnt/f/NSL/src/NSL/FermionMatrix/Impl/hubbardExp.tpp:46:52:   required from ‘NSL::Tensor<Type> NSL::FermionMatrix::FermionMatrixHubbardExp<Type>::Mdagger(const NSL::Tensor<Type>&) [with Type = c10::complex<float>]’
/mnt/f/NSL/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:165:59:   required from ‘void test_fermionMatrixHubbardExp_Mdagger(NSL::size_t, NSL::size_t) [with Type = float; NSL::size_t = long int]’
/mnt/f/NSL/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:41:51:   required from ‘void C_A_T_C_H_T_E_M_P_L_A_T_E_T_E_S_T_F_U_N_C_7() [with TestType = float]’
/mnt/f/NSL/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:37:1:   required from ‘void {anonymous}::ns_C_A_T_C_H_T_E_M_P_L_A_T_E_T_E_S_T_8::reg_test({anonymous}::ns_C_A_T_C_H_T_E_M_P_L_A_T_E_T_E_S_T_8::TypeList<Type>, Catch::NameAndTags) [with Type = float]’
/mnt/f/NSL/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:37:1:   required from ‘{anonymous}::ns_C_A_T_C_H_T_E_M_P_L_A_T_E_T_E_S_T_8::C_A_T_C_H_T_E_M_P_L_A_T_E_T_E_S_T_8<Types>::C_A_T_C_H_T_E_M_P_L_A_T_E_T_E_S_T_8() [with Types = {{anonymous}::ns_C_A_T_C_H_T_E_M_P_L_A_T_E_T_E_S_T_8::TypeList<float>, {anonymous}::ns_C_A_T_C_H_T_E_M_P_L_A_T_E_T_E_S_T_8::TypeList<double>}]’
/mnt/f/NSL/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:37:1:   required from here
/mnt/f/NSL/src/NSL/Tensor/Impl/shift.tpp:32:57: error: ‘class NSL::TensorImpl::TensorShift<c10::complex<float> >’ has no member named ‘shape’
   32 |             this->data_.slice(/*dim=*/0,/*start=*/this->shape(0)-shift,/*end=*/this->shape(0),/*step=*/1)*=boundary;
      |                                                   ~~~~~~^~~~~
/mnt/f/NSL/src/NSL/Tensor/Impl/shift.tpp:32:86: error: ‘class NSL::TensorImpl::TensorShift<c10::complex<float> >’ has no member named ‘shape’
   32 |             this->data_.slice(/*dim=*/0,/*start=*/this->shape(0)-shift,/*end=*/this->shape(0),/*step=*/1)*=boundary;
      |                                                                                ~~~~~~^~~~~
/mnt/f/NSL/src/NSL/Tensor/Impl/shift.tpp: In instantiation of ‘NSL::Tensor<Type> NSL::TensorImpl::TensorShift<Type>::shift(const size_t&, const Type&) [with Type = c10::complex<double>; NSL::size_t = long int]’:
/mnt/f/NSL/src/NSL/LinAlg/shift.tpp:23:43:   required from ‘NSL::Tensor<Type> NSL::LinAlg::shift(const NSL::Tensor<Type>&, const size_t&, const Type&) [with Type = c10::complex<double>; NSL::size_t = long int]’
/mnt/f/NSL/src/NSL/FermionMatrix/Impl/hubbardExp.tpp:46:52:   required from ‘NSL::Tensor<Type> NSL::FermionMatrix::FermionMatrixHubbardExp<Type>::Mdagger(const NSL::Tensor<Type>&) [with Type = c10::complex<double>]’
/mnt/f/NSL/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:165:59:   required from ‘void test_fermionMatrixHubbardExp_Mdagger(NSL::size_t, NSL::size_t) [with Type = double; NSL::size_t = long int]’
/mnt/f/NSL/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:41:51:   required from ‘void C_A_T_C_H_T_E_M_P_L_A_T_E_T_E_S_T_F_U_N_C_7() [with TestType = double]’
/mnt/f/NSL/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:37:1:   required from ‘void {anonymous}::ns_C_A_T_C_H_T_E_M_P_L_A_T_E_T_E_S_T_8::reg_test({anonymous}::ns_C_A_T_C_H_T_E_M_P_L_A_T_E_T_E_S_T_8::TypeList<Type>, Catch::NameAndTags) [with Type = double]’
/mnt/f/NSL/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:37:1:   required from ‘{anonymous}::ns_C_A_T_C_H_T_E_M_P_L_A_T_E_T_E_S_T_8::C_A_T_C_H_T_E_M_P_L_A_T_E_T_E_S_T_8<Types>::C_A_T_C_H_T_E_M_P_L_A_T_E_T_E_S_T_8() [with Types = {{anonymous}::ns_C_A_T_C_H_T_E_M_P_L_A_T_E_T_E_S_T_8::TypeList<float>, {anonymous}::ns_C_A_T_C_H_T_E_M_P_L_A_T_E_T_E_S_T_8::TypeList<double>}]’
/mnt/f/NSL/Tests/FermionMatrix/test_fermionMatrixHubbardExp.cpp:37:1:   required from here
/mnt/f/NSL/src/NSL/Tensor/Impl/shift.tpp:32:57: error: ‘class NSL::TensorImpl::TensorShift<c10::complex<double> >’ has no member named ‘shape’
   32 |             this->data_.slice(/*dim=*/0,/*start=*/this->shape(0)-shift,/*end=*/this->shape(0),/*step=*/1)*=boundary;
      |                                                   ~~~~~~^~~~~
/mnt/f/NSL/src/NSL/Tensor/Impl/shift.tpp:32:86: error: ‘class NSL::TensorImpl::TensorShift<c10::complex<double> >’ has no member named ‘shape’
   32 |             this->data_.slice(/*dim=*/0,/*start=*/this->shape(0)-shift,/*end=*/this->shape(0),/*step=*/1)*=boundary;
      |                                                                                ~~~~~~^~~~~```

evanberkowitz committed:
Merge branch 'devel' into feature/fermionMatrix/HubbardExp

evanberkowitz committed:
Give logdetM_* tests meaningful names.

evanberkowitz committed:
Rename size_0 --> nt and size_1 --> nx

evanberkowitz committed:
Put beta as a default argument for tests.

Then
 - we can change it!
 - no magic constants in the construction of M.

evanberkowitz committed:
Prevent Torch runtime warning during adjacency matrix construction

evanberkowitz committed:
COMPLEX_NSL_TEST_CASE tests complex types only

evanberkowitz committed:
In HubbardExp constructor beta is Type, not double

evanberkowitz committed:
Add Complete<complex<...>>; see #70

evanberkowitz committed:
Add NSL::to_string to stringify complex numbers

evanberkowitz committed:
Remove explicit Ring from tests.

 - Simplify printing with NSL::to_string.
 - Put the LatticeType as a template parameter for all tests.
 - Remove some loops in favor of implicit Tensor manipulations.
 - Add nx and nt reporting to INFO.

evanberkowitz committed:
Test self-consistency for HubbardExp MM† and M†M

evanberkowitz committed:
FermionMatrix tests require nonzero phi

Otherwise the tests will pass even in some situations where they are
supposed to fail.

evanberkowitz commented:
We should consider merging #71 into this PR before proceeding any further.

evanberkowitz committed:
Generic M_dense

evanberkowitz committed:
Bootstrap test correctness

It's too fragile to reimplement all the matrix-vector calls.

Instead, we test that
 - the application of M is correct,
 - we can get a dense representation of M by repeated application of M
 - M† is correct by checking against an obvious matrix-vector multiply using the dense representation
 - MdaggerM = Mdagger • M
 - MMdagger = M       • Mdagger.

That covers ALL of our bases and we don't have to copy the
implementation of M into the test---we test some for correctness
and then test the rest using the logical relationships.

evanberkowitz commented:
> We should consider merging #71 into this PR before proceeding any further.

Done; #71 is now merged into this PR.

evanberkowitz committed:
Add to exponential discretization docs.

evanberkowitz committed:
Add citations to documentation.

evanberkowitz committed:
Add references.

evanberkowitz committed:
Prevent doubling of documentation of F_

evanberkowitz committed:
Sign convention in docs

evanberkowitz committed:
Document the sparse logic of M and M† and implement

Still to do:  the signs in the shift?

But, Mdagger tests pass (usually).

evanberkowitz committed:
Hubbard Exp MM† and M†M: derive + implement

evanberkowitz commented:
I think I have debugged the M, M†, MM† and M†M implementations.  It would be good for @luutom, @Marcel-Rodekamp, and @keshvituteja to really go through my allegations very carefully.

The one thing I am uncertain of is the directions of the shifts.

evanberkowitz committed:
Remove extraneous declaration.

evanberkowitz committed:
Simplify Hubbard Exp logdet

evanberkowitz commented:
Notably, the M†, MM†, and M†M tests now pass for `complex<double>`.  The sparse/dense comparison for `M` passes.  For the `complex<double>` test for M I get (for example)
```
test cases:   1 |   0 passed |  1 failed
assertions: 245 | 196 passed | 49 failed
```
so this may be correct; it might not be.  I purposefully haven't adjusted the test of M to match my implementation for two reasons:
 - We shouldn't until people carefully check my implementation.
 - We shouldn't until we actually figure out the correct direction of the shift!  I'm still mixed up by this!

evanberkowitz committed:
Implement and use almost_equal for NSL::Tensors

evanberkowitz committed:
Forgot to commit removal of epsilon from implementations.

Marcel-Rodekamp committed:
removed todo as it is done

Marcel-Rodekamp committed:
Fixed little typo in in code documentation

Chrisgant committed:
Fixed Typo

Chrisgant committed:
Fixed Typo

Marcel-Rodekamp committed:
Change to a more digestible (and now correct) way of comparing digits in base 10

Marcel-Rodekamp committed:
Bugfix in first test to 'different' implementation

Marcel-Rodekamp committed:
Adjusted tests to change of almost_equal

evanberkowitz commented:
To do physics we still need to handle particles / holes.  But now I think this FermionMatrix (which is for Hubbard particles) is right up to
 - no chemical potential µ
 - no holes
 - we should add a test of the actual dense det to check that the values come out right
